src/buffer.rs

```diff
@@ -577,10 +577,19 @@ impl PyBuffer {
     }
 
     pub fn release(self, _py: Python) {
+        // First move self into a ManuallyDrop, so that PyBuffer::drop will
+        // never be called. (It would acquire the GIL and call PyBuffer_Release
+        // again.)
+        let mut mdself = mem::ManuallyDrop::new(self);
         unsafe {
-            let ptr = &*self.0 as *const ffi::Py_buffer as *mut ffi::Py_buffer;
-            ffi::PyBuffer_Release(ptr)
-        };
+            // Next, make the actual PyBuffer_Release call.
+            ffi::PyBuffer_Release(&mut *mdself.0);
+
+            // Finally, drop the contained Pin<Box<_>> in place, to free the
+            // Box memory.
+            let inner: *mut Pin<Box<ffi::Py_buffer>> = &mut mdself.0;
+            ptr::drop_in_place(inner);
```

Comment:
```suggestion
            ptr::drop_in_place(&mut mdself.0);
```

Reply:
I did it this way because I'm a little bit afraid of (future) Deref magic. For example, it looks like there's some Deref magic happening in the line `ffi::PyBuffer_Release(&mut *mdself.0)`. I'm worried that someone else might change this type in the future, and if so I _want_ this line to break if the type changes. It's too easy for `drop_in_place` to accidentally become a no-op, for example if you pass it a raw `*mut T`.

Reply:
OK, let's leave it as is :+1: